src/main/java/com/wavefront/generators/MainGenerator.java

```diff
@@ -0,0 +1,45 @@
+package com.wavefront.generators;
+
+import com.wavefront.DataQueue;
+import com.wavefront.datastructures.Trace;
+import com.wavefront.helpers.Statistics;
+
+import org.apache.commons.lang3.NotImplementedException;
+
+import java.util.Random;
+
+import javax.annotation.Nonnull;
+
+public abstract class MainGenerator implements Runnable {
+  protected static final Random RANDOM = new Random(System.currentTimeMillis());
+  protected static final int SLEEP_DELAY_SECONDS = 5;
+  protected static final int SLEEP_DELAY_MILLIS = SLEEP_DELAY_SECONDS * 1000;
+  public final Statistics statistics = new Statistics();
+  @Nonnull
+  protected final DataQueue dataQueue;
+
+  protected MainGenerator(@Nonnull DataQueue dataQueue) {
+    this.dataQueue = dataQueue;
+  }
+
+  /**
+   * Generates traces for saving to file (without ingestion specific time delays).
+   */
+  public abstract void generateForFile();
+
+  /**
+   * Returns statistics about the generated traces.
+   */
+  public abstract Statistics getStatistics();
+
+  /**
+   * Generate trace for a given trace type.
+   *
+   * @param startMillis Start time of the trace (millis).
+   * @return Generated trace.
+   */
+  protected Trace generateTrace(long startMillis) {
```

Comment:
I guess the place of this method is in the TraceGenerator class. As far as I understand ReIngestGenerator doesn't override this method.